### PR TITLE
Add option for bootstrapped confidence intervals around calculated rewards

### LINKED ISSUE #67: Confidence intervals on OPE estimates
Can do this using bootstrapping.

### PR DESCRIPTION
**Pull request by fergaljoconnor**
Resolves banditml/offline-policy-evaluation#67

### CODE REVIEW

File: ope/methods/doubly_robust.py [hunk 1]
```diff
@@ -35,7 +49,7 @@ def evaluate(df: pd.DataFrame, action_prob_function: Callable) -> Dict[str, floa
 
         # then compute the right hand term, which is similar to IPS
         logged_action = row["action"]
-        new_action_probability = action_probabilities[row["action"]]
+        new_action_probability = action_probabilities[logged_action]
```
Comment (econti) [inline ope/methods/doubly_robust.py:52]: thanks, good catch.

File: ope/methods/inverse_propensity_scoring.py [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
     """
+    return evaluation_and_bootstrap_metrics(
+        df,
+        lambda dataframe: evaluate_raw(dataframe, action_prob_function),
+        num_bootstrap_samples=num_bootstrap_samples,
+    )
+
 
+def evaluate_raw(df: pd.DataFrame, action_prob_function: Callable) -> Dict[str, float]:
+    """Evaluate a policy on a dataset with a pre-trained model"""
```
Comment (econti) [inline ope/methods/inverse_propensity_scoring.py:23]: nit: no pre-trained model here

File: ope/sampling/bootstrap.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+from typing import Callable, Dict, Iterable, Union
+
+import pandas as pd
+from pandas import DataFrame
```
Comment (econti) [inline ope/sampling/bootstrap.py:4]: nit: delete and use `pd.`

File: ope/sampling/bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+from .metric import confidence_interval
+
+
+def evaluation_and_bootstrap_metrics(
+    data: DataFrame,
```
Comment (econti) [inline ope/sampling/bootstrap.py:10]: nit: `pd.DataFrame` to be consistent

File: ope/sampling/metric.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    def get_confidence_interval(series: pd.Series) -> Dict[str, Union[str, float]]:
+        mean = series.mean()
+        standard_dev = series.std()
+        return {
+            "name": f"confidence_interval_{p_value}",
+            "value": [mean - 1.96 * standard_dev, mean + 1.96 * standard_dev],
+        }
+
+    return get_confidence_interval
```
Comment (econti) [inline ope/sampling/metric.py:17]: I get why you are returning a `Callable`, but to keep things simple, could we just have a `compute_confidence_interval(series, p_value) -> Dict` function that takes in the series and a p_value and returns the confidence interval? I like to keep the readability of less fancy code.

File: ope/sampling/metric.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+def confidence_interval(p_value: float) -> Callable:
+    """Returns a callable which evaluates a confidence interval on a series."""
+
+    def get_confidence_interval(series: pd.Series) -> Dict[str, Union[str, float]]:
+        mean = series.mean()
+        standard_dev = series.std()
+        return {
+            "name": f"confidence_interval_{p_value}",
+            "value": [mean - 1.96 * standard_dev, mean + 1.96 * standard_dev],
```
Comment (econti) [inline ope/sampling/metric.py:14]: nit: pull out 1.96 as a constant somewhere

**Review by econti — COMMENTED** [review]
Thanks for all the work, overall looks awesome! Had just a few nits and one larger request. Reducing the amount of functional programming here would be preferred to keep things easy / simple to understand. Would be much appreciated if you can make a few tweaks to do that. If you don't have the time, no worries, I can take over the PR and take care of it.

File: ope/sampling/metric.py (hunk 6 above)
Comment (fergaljoconnor) [reply ope/sampling/metric.py:14]: Apologies, no idea how the magic number version is still in here, this was the quick and dirty one I was using for early testing. Had a version that uses scipy.norm if you're comfortable having scipy as a dependency (It's a pretty stable dependency of scikit-learn so it shouldn't pull the rug out from under anything for now unless you want to feature-gate the scikit-learn stuff at any point).

File: ope/sampling/metric.py (hunk 5 above)
Comment (fergaljoconnor) [reply ope/sampling/metric.py:17]: Perfect, no hassle

File: ope/sampling/metric.py (hunk 6 above)
Comment (econti) [reply ope/sampling/metric.py:14]: No problem. Prefer to leave Scipy out for now!

File: ope/sampling/metric.py (hunk 5 above)
Comment (econti) [reply ope/sampling/metric.py:17]: Thank you!

**Comment by fergaljoconnor** [comment]
Cool. Just tossing around ways to ditch the functional baggage in my head. The bootstrap code probably needs to be completely ignorant of any concerns apart from sampling the data and performing some operation on it. I'm using the callable passed in as an evaluator for that operation right now, but one way of getting less functional would be to have it instead be passed some object that has an evaluate function with an interface something like:

```python
class PolicyEvaluator:
    def __init__(self, ...args, ...kwargs):
        # kwargs and args are whatever you need to generate 
        # an evaluator for a specific policy. For example one of 
        # the machine learning-based methods would need the data to train

    def evaluate(self, df) -> DictWithModelEvaluation:
        # Same interface as the standard evaluate function
```

At this point you can pass one of these into the bootstrapping code and it'll use it to generate the outputs and combine them. No lambdas and a pretty big chunk of boilerplate (I think you need a subclass of the above for every different evaluation method), but the code should be more familiar to anyone who has ever called a class constructor in Python.

That being said, this still might be a lot of extra machinery for what the code needs. If you have any cleaner ideas let me know. If not I can with something like the above.

I think the above and the confidence intervals would be the main functional choke points, but similarly let me know if there are any other parts of the code that stick out.

**Comment by econti** [comment]
Cool, let me think about it a bit.

**Comment by econti** [comment]
Hey @fergaljoconnor -- after thinking through it I went ahead and implemented what I was thinking here:
https://github.com/banditml/offline-policy-evaluation/pull/72

I mainly copied what you did, but made the code less functional 😬 for readability. Sorry for all the back and forth nits, your PR was awesome, I just find dumb code easier to read sometimes.

Let me know if the PR looks good and I'll merge it. Thank you again for all the help and contributing great ideas.

**Comment by fergaljoconnor** [comment]
No objections here, the new PR looks very clear. 👍
